src/net.cpp

```diff
@@ -3196,30 +3380,32 @@ void CNode::EndMessage() UNLOCK_FUNCTION(cs_vSend)
         strcmp(strCommand, NetMsgType::VERACK) != 0 && strcmp(strCommand, NetMsgType::INV) != 0)
     {
         nActivityBytes += nSize;
+    }
 
-        // If the message is a priority message then move to the front of the deque
-        if (strcmp(strCommand, NetMsgType::GET_XTHIN) == 0 || strcmp(strCommand, NetMsgType::XTHINBLOCK) == 0 ||
-            strcmp(strCommand, NetMsgType::GET_THIN) == 0 || strcmp(strCommand, NetMsgType::THINBLOCK) == 0 ||
-            strcmp(strCommand, NetMsgType::XBLOCKTX) == 0 || strcmp(strCommand, NetMsgType::GET_XBLOCKTX) == 0 ||
-            strcmp(strCommand, NetMsgType::GET_GRAPHENE) == 0 || strcmp(strCommand, NetMsgType::GRAPHENEBLOCK) == 0 ||
-            strcmp(strCommand, NetMsgType::GRAPHENETX) == 0 || strcmp(strCommand, NetMsgType::GET_GRAPHENETX) == 0)
-        {
-            it = vSendMsg.insert(vSendMsg.begin(), CSerializeData());
-            LOG(THIN, "Send Queue: pushed %s to the front of the queue\n", strCommand);
-        }
-        else
-            it = vSendMsg.insert(vSendMsg.end(), CSerializeData());
+
+    // If the message is a priority message then move to the front of the deque and add
+    // a CNodeRef entry into the vPrioritySendQ. Now that that message is at the front of the queue
+    // all we have to do is emtpy this first item from the vSendMsg queue when we SendMessages().
+    if (IsPriorityMsg(strCommand))
+    {
+        it = vSendMsg.insert(vSendMsg.begin(), CSerializeData());
+
+        // Do this here because we may swap the iterator later.
+        ssSend.GetAndClear(*it);
+        nSendSize.fetch_add((*it).size());
+
+        // Add to the priority queue and position the message by swapping the iterators.
+        PrioritizeSendMsg(this);
+        LOG(THIN | GRAPHENE | CMPCT, "Send Queue: pushed %s to the front of the queue, peer(%d)\n", strCommand,
+            this->GetId());
     }
     else
+    {
         it = vSendMsg.insert(vSendMsg.end(), CSerializeData());
-    // BU: end
-
-    ssSend.GetAndClear(*it);
-    nSendSize.fetch_add((*it).size());
+        ssSend.GetAndClear(*it);
+        nSendSize.fetch_add((*it).size());
+    }
 
-    // If write queue empty, attempt "optimistic write"
-    if (it == vSendMsg.begin())
-        SocketSendData(this);
```

Comment:
I added this code as part of gigaperf because it made a big difference in minimizing message latency.  Perhaps we've fixed whatever was causing the latency though... maybe comment out with a note and we can try it on and off in gigaperf test 2

Reply:
I think  you might be right...I'll revert for now and we can retest again later.